relayer/relays/beacon/header/header.go

```diff
@@ -0,0 +1,262 @@
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	log "github.com/sirupsen/logrus"
+	"github.com/snowfork/go-substrate-rpc-client/v4/types"
+	"github.com/snowfork/snowbridge/relayer/relays/beacon/cache"
+	"github.com/snowfork/snowbridge/relayer/relays/beacon/header/syncer"
+	"github.com/snowfork/snowbridge/relayer/relays/beacon/header/syncer/scale"
+	"github.com/snowfork/snowbridge/relayer/relays/beacon/writer"
+)
+
+var ErrFinalizedHeaderUnchanged = errors.New("finalized header unchanged")
+
+type Header struct {
+	cache  *cache.BeaconCache
+	writer *writer.ParachainWriter
+	syncer *syncer.Syncer
+}
+
+func New(writer *writer.ParachainWriter, beaconEndpoint string, slotsInEpoch uint64, epochsPerSyncCommitteePeriod uint64) Header {
+	return Header{
+		cache:  cache.New(),
+		writer: writer,
+		syncer: syncer.New(beaconEndpoint, slotsInEpoch, epochsPerSyncCommitteePeriod),
+	}
+}
+
+func (h *Header) Sync(ctx context.Context) (<-chan uint64, <-chan uint64, error) {
+	latestSyncedPeriod, err := h.writer.GetLastSyncedSyncCommitteePeriod()
+	if err != nil {
+		return nil, nil, fmt.Errorf("fetch last sync commitee: %w", err)
+	}
+
+	h.cache.SetLastSyncedSyncCommitteePeriod(latestSyncedPeriod)
+
+	log.WithField("period", latestSyncedPeriod).Info("set cache: last beacon synced sync committee period")
+
+	periodsToSync, err := h.syncer.GetSyncPeriodsToFetch(latestSyncedPeriod)
+	if err != nil {
+		return nil, nil, fmt.Errorf("check sync committee periods to be fetched: %w", err)
+	}
+
+	log.WithFields(log.Fields{
+		"periods": periodsToSync,
+	}).Info("sync committee periods to be synced")
+
+	for _, period := range periodsToSync {
+		err := h.SyncCommitteePeriodUpdate(ctx, period)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	lastFinalizedHeader, err := h.writer.GetLastStoredFinalizedHeader()
+	if err != nil {
+		return nil, nil, fmt.Errorf("fetch last finalized header: %w", err)
+	}
+
+	lastFinalizedSlot, err := h.writer.GetLastStoredFinalizedHeaderSlot()
+	if err != nil {
+		return nil, nil, fmt.Errorf("fetch last finalized header slot: %w", err)
+	}
+
+	h.cache.FinalizedHeaders = append(h.cache.FinalizedHeaders, lastFinalizedHeader)
+
+	log.WithFields(log.Fields{
+		"hash": lastFinalizedHeader,
+		"slot": lastFinalizedSlot,
+	}).Info("set cache: last finalized header")
+
+	log.Info("starting to sync finalized headers")
+
+	ticker := time.NewTicker(time.Second * 20)
+	done := make(chan bool)
+
+	basicChannel := make(chan uint64)
+	incentivizedChannel := make(chan uint64)
+
+	go func() {
+		for {
+			lastBlockNumber, err := h.SyncHeaders(ctx)
+			switch {
+			case errors.Is(err, ErrFinalizedHeaderUnchanged):
+				log.WithField("finalized_header", h.cache.LastFinalizedHeader()).Info("finalized header unchanged")
+			case err != nil:
+				log.WithError(err).Error("error while syncing headers")
+
+				return
+			default:
+				basicChannel <- lastBlockNumber
+				incentivizedChannel <- lastBlockNumber
```

Comment:
I've made 2 separate channels for the basic and incentivized channel. I read up on patterns of 1 producer, 2 readers, and it looks like if you want to go that route you need to crate a publish and subscribe mechanism. That felt overkill for just the 2 channels so I left it like this.